### PR TITLE
Use the instance id as the volume index

### LINKED ISSUE #2376: Find a way to assign volume IDs to instances without EBS
### Context

https://github.com/AlexsLemonade/refinebio/issues/2254

### Problem or idea

We stopped creating EBS volumes now that we do all of the processing on the SSDs, but now we can only run one instance at once because instances are always assigned volume index 0.

### Solution or next step

Brainstorm ways to assign volume indices to instances. Maybe we can query nomad for active instances?

### PR DESCRIPTION
**Pull request by wvauclain**
## Issue Number

Fixes #2376

## Purpose/Implementation Notes

Use the instance id as the volume index, and dynamically create new nomad jobs when an instance launches instead of pre-allocating a certain number of volume indices.

I still need to add code to the janitor to clean up extra jobs once an instance goes down.

## Types of changes

What types of changes does your code introduce?

- New feature (non-breaking change which adds functionality)

## Functional tests

I ran on a devstack and watched it successfully assign samples to an instance and process samples.

After I add the foreman code I should deploy and test it one more time.

## Checklist

_Put an `x` in the boxes that apply._

- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Screenshots

<details>
<summary>The output of `nomad status`</summary>

```
ID                                                                 Type                 Priority  Status   Submit Date
AFFY_TO_PCL_i-0e7054c600dfc3dd3_12288                              batch/parameterized  50        running  2020-08-03T18:06:17Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_16384                              batch/parameterized  50        running  2020-08-03T18:06:17Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048                               batch/parameterized  50        running  2020-08-03T18:06:17Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478024-1681ac0c  batch                50        dead     2020-08-03T18:07:04Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478024-92972fde  batch                50        dead     2020-08-03T18:07:04Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478024-9e83ee08  batch                50        dead     2020-08-03T18:07:04Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478024-e4f42dc2  batch                50        dead     2020-08-03T18:07:04Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478024-eba9c5ec  batch                50        dead     2020-08-03T18:07:04Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478025-6a666e43  batch                50        dead     2020-08-03T18:07:05Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478025-83d6bfce  batch                50        dead     2020-08-03T18:07:05Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478025-c2fdba52  batch                50        dead     2020-08-03T18:07:05Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478025-e1a657dc  batch                50        dead     2020-08-03T18:07:05Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478025-f726c063  batch                50        dead     2020-08-03T18:07:05Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478033-337d717f  batch                50        dead     2020-08-03T18:07:13Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478033-650263d4  batch                50        dead     2020-08-03T18:07:13Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478033-86f39657  batch                50        dead     2020-08-03T18:07:13Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478033-bd336713  batch                50        dead     2020-08-03T18:07:13Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478033-c35b2765  batch                50        dead     2020-08-03T18:07:13Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478034-1aa89c16  batch                50        dead     2020-08-03T18:07:14Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478034-1eafc7da  batch                50        dead     2020-08-03T18:07:14Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478034-69110036  batch                50        dead     2020-08-03T18:07:14Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478034-b40528e0  batch                50        dead     2020-08-03T18:07:14Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478034-e962ce89  batch                50        dead     2020-08-03T18:07:14Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478048-1783f9a7  batch                50        dead     2020-08-03T18:07:28Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478048-2a4ecda7  batch                50        dead     2020-08-03T18:07:28Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478048-65c0f249  batch                50        dead     2020-08-03T18:07:28Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478048-7fe9b7fc  batch                50        dead     2020-08-03T18:07:28Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478048-8cce43b2  batch                50        dead     2020-08-03T18:07:28Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478048-9596458b  batch                50        dead     2020-08-03T18:07:28Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478048-a2cda93c  batch                50        dead     2020-08-03T18:07:28Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478048-b727c3a5  batch                50        dead     2020-08-03T18:07:28Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478048-da566ff5  batch                50        dead     2020-08-03T18:07:28Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_2048/dispatch-1596478048-ff93308b  batch                50        dead     2020-08-03T18:07:28Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_3072                               batch/parameterized  50        running  2020-08-03T18:06:17Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_32768                              batch/parameterized  50        running  2020-08-03T18:06:17Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_4096                               batch/parameterized  50        running  2020-08-03T18:06:17Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_65536                              batch/parameterized  50        running  2020-08-03T18:06:18Z
AFFY_TO_PCL_i-0e7054c600dfc3dd3_8192                               batch/parameterized  50        running  2020-08-03T18:06:18Z
CREATE_COMPENDIA                                                   batch/parameterized  80        running  2020-08-03T17:28:19Z
CREATE_QN_TARGET                                                   batch/parameterized  80        running  2020-08-03T17:28:19Z
CREATE_QUANTPENDIA                                                 batch/parameterized  80        running  2020-08-03T17:28:19Z
DOWNLOADER_i-0e7054c600dfc3dd3_1024                                batch/parameterized  40        running  2020-08-03T18:06:18Z
DOWNLOADER_i-0e7054c600dfc3dd3_1024/dispatch-1596477991-4435dbd4   batch                40        dead     2020-08-03T18:06:31Z
DOWNLOADER_i-0e7054c600dfc3dd3_1024/dispatch-1596477991-62c6a81a   batch                40        dead     2020-08-03T18:06:31Z
DOWNLOADER_i-0e7054c600dfc3dd3_1024/dispatch-1596477991-64b3a5a0   batch                40        dead     2020-08-03T18:06:31Z
DOWNLOADER_i-0e7054c600dfc3dd3_1024/dispatch-1596477991-798946c1   batch                40        dead     2020-08-03T18:06:31Z
DOWNLOADER_i-0e7054c600dfc3dd3_1024/dispatch-1596477991-a5d27907   batch                40        dead     2020-08-03T18:06:31Z
DOWNLOADER_i-0e7054c600dfc3dd3_1024/dispatch-1596477991-bb4d1f51   batch                40        dead     2020-08-03T18:06:31Z
DOWNLOADER_i-0e7054c600dfc3dd3_1024/dispatch-1596477991-ddc8ba51   batch                40        dead     2020-08-03T18:06:31Z
DOWNLOADER_i-0e7054c600dfc3dd3_1024/dispatch-1596477991-e0ed8a38   batch                40        dead     2020-08-03T18:06:31Z
DOWNLOADER_i-0e7054c600dfc3dd3_1024/dispatch-1596477991-f2a461d6   batch                40        dead     2020-08-03T18:06:31Z
DOWNLOADER_i-0e7054c600dfc3dd3_1024/dispatch-1596477991-f976e799   batch                40        dead     2020-08-03T18:06:31Z
DOWNLOADER_i-0e7054c600dfc3dd3_1024/dispatch-1596478175-ef53dab8   batch                40        dead     2020-08-03T18:09:35Z
DOWNLOADER_i-0e7054c600dfc3dd3_16384                               batch/parameterized  40        running  2020-08-03T18:06:18Z
DOWNLOADER_i-0e7054c600dfc3dd3_4096                                batch/parameterized  40        running  2020-08-03T18:06:18Z
ILLUMINA_TO_PCL_i-0e7054c600dfc3dd3_12288                          batch/parameterized  50        running  2020-08-03T18:06:18Z
ILLUMINA_TO_PCL_i-0e7054c600dfc3dd3_16384                          batch/parameterized  50        running  2020-08-03T18:06:18Z
ILLUMINA_TO_PCL_i-0e7054c600dfc3dd3_2048                           batch/parameterized  50        running  2020-08-03T18:06:18Z
ILLUMINA_TO_PCL_i-0e7054c600dfc3dd3_3072                           batch/parameterized  50        running  2020-08-03T18:06:18Z
ILLUMINA_TO_PCL_i-0e7054c600dfc3dd3_32768                          batch/parameterized  50        running  2020-08-03T18:06:18Z
ILLUMINA_TO_PCL_i-0e7054c600dfc3dd3_4096                           batch/parameterized  50        running  2020-08-03T18:06:18Z
ILLUMINA_TO_PCL_i-0e7054c600dfc3dd3_65536                          batch/parameterized  50        running  2020-08-03T18:06:18Z
ILLUMINA_TO_PCL_i-0e7054c600dfc3dd3_8192                           batch/parameterized  50        running  2020-08-03T18:06:18Z
JANITOR_i-0e7054c600dfc3dd3_12288                                  batch/parameterized  75        running  2020-08-03T18:06:18Z
JANITOR_i-0e7054c600dfc3dd3_16384                                  batch/parameterized  75        running  2020-08-03T18:06:18Z
JANITOR_i-0e7054c600dfc3dd3_2048                                   batch/parameterized  75        running  2020-08-03T18:06:18Z
JANITOR_i-0e7054c600dfc3dd3_3072                                   batch/parameterized  75        running  2020-08-03T18:06:18Z
JANITOR_i-0e7054c600dfc3dd3_32768                                  batch/parameterized  75        running  2020-08-03T18:06:18Z
JANITOR_i-0e7054c600dfc3dd3_4096                                   batch/parameterized  75        running  2020-08-03T18:06:19Z
JANITOR_i-0e7054c600dfc3dd3_65536                                  batch/parameterized  75        running  2020-08-03T18:06:19Z
JANITOR_i-0e7054c600dfc3dd3_8192                                   batch/parameterized  75        running  2020-08-03T18:06:19Z
NO_OP_i-0e7054c600dfc3dd3_12288                                    batch/parameterized  50        running  2020-08-03T18:06:19Z
NO_OP_i-0e7054c600dfc3dd3_16384                                    batch/parameterized  50        running  2020-08-03T18:06:19Z
NO_OP_i-0e7054c600dfc3dd3_2048                                     batch/parameterized  50        running  2020-08-03T18:06:19Z
NO_OP_i-0e7054c600dfc3dd3_2048/dispatch-1596478183-0ec1d3d9        batch                50        dead     2020-08-03T18:09:43Z
NO_OP_i-0e7054c600dfc3dd3_2048/dispatch-1596478183-2cc4241e        batch                50        dead     2020-08-03T18:09:43Z
NO_OP_i-0e7054c600dfc3dd3_2048/dispatch-1596478183-68cfa0af        batch                50        dead     2020-08-03T18:09:43Z
NO_OP_i-0e7054c600dfc3dd3_2048/dispatch-1596478183-f8b48497        batch                50        dead     2020-08-03T18:09:43Z
NO_OP_i-0e7054c600dfc3dd3_2048/dispatch-1596478184-18133416        batch                50        dead     2020-08-03T18:09:44Z
NO_OP_i-0e7054c600dfc3dd3_2048/dispatch-1596478184-7428037b        batch                50        dead     2020-08-03T18:09:44Z
NO_OP_i-0e7054c600dfc3dd3_2048/dispatch-1596478184-766c3d49        batch                50        dead     2020-08-03T18:09:44Z
NO_OP_i-0e7054c600dfc3dd3_2048/dispatch-1596478184-786af29f        batch                50        dead     2020-08-03T18:09:44Z
NO_OP_i-0e7054c600dfc3dd3_2048/dispatch-1596478184-a61ddc17        batch                50        dead     2020-08-03T18:09:44Z
NO_OP_i-0e7054c600dfc3dd3_2048/dispatch-1596478184-b3857e00        batch                50        dead     2020-08-03T18:09:44Z
NO_OP_i-0e7054c600dfc3dd3_2048/dispatch-1596478184-cc079c48        batch                50        dead     2020-08-03T18:09:44Z
NO_OP_i-0e7054c600dfc3dd3_2048/dispatch-1596478184-d9e7d132        batch                50        dead     2020-08-03T18:09:44Z
NO_OP_i-0e7054c600dfc3dd3_3072                                     batch/parameterized  50        running  2020-08-03T18:06:19Z
NO_OP_i-0e7054c600dfc3dd3_32768                                    batch/parameterized  50        running  2020-08-03T18:06:19Z
NO_OP_i-0e7054c600dfc3dd3_4096                                     batch/parameterized  50        running  2020-08-03T18:06:19Z
NO_OP_i-0e7054c600dfc3dd3_65536                                    batch/parameterized  50        running  2020-08-03T18:06:19Z
NO_OP_i-0e7054c600dfc3dd3_8192                                     batch/parameterized  50        running  2020-08-03T18:06:19Z
QN_DISPATCHER                                                      batch/parameterized  90        running  2020-08-03T18:06:19Z
QN_REFERENCE                                                       batch/parameterized  80        running  2020-08-03T18:06:19Z
SALMON_i-0e7054c600dfc3dd3_12288                                   batch/parameterized  50        running  2020-08-03T18:06:20Z
SALMON_i-0e7054c600dfc3dd3_16384                                   batch/parameterized  50        running  2020-08-03T18:06:20Z
SALMON_i-0e7054c600dfc3dd3_2048                                    batch/parameterized  50        running  2020-08-03T18:06:20Z
SALMON_i-0e7054c600dfc3dd3_3072                                    batch/parameterized  50        running  2020-08-03T18:06:20Z
SALMON_i-0e7054c600dfc3dd3_32768                                   batch/parameterized  50        running  2020-08-03T18:06:20Z
SALMON_i-0e7054c600dfc3dd3_4096                                    batch/parameterized  50        running  2020-08-03T18:06:20Z
SALMON_i-0e7054c600dfc3dd3_65536                                   batch/parameterized  50        running  2020-08-03T18:06:20Z
SALMON_i-0e7054c600dfc3dd3_8192                                    batch/parameterized  50        running  2020-08-03T18:06:20Z
SEQUENTIAL_SURVEYOR_16384                                          batch/parameterized  30        running  2020-08-03T17:28:19Z
SEQUENTIAL_SURVEYOR_256                                            batch/parameterized  30        running  2020-08-03T17:28:19Z
SEQUENTIAL_SURVEYOR_4096                                           batch/parameterized  30        running  2020-08-03T17:28:19Z
SMASHER                                                            batch/parameterized  75        running  2020-08-03T18:06:20Z
SURVEYOR_16384                                                     batch/parameterized  30        running  2020-08-03T17:28:19Z
SURVEYOR_256                                                       batch/parameterized  30        running  2020-08-03T17:28:19Z
SURVEYOR_4096                                                      batch/parameterized  30        running  2020-08-03T17:28:19Z
SURVEYOR_DISPATCHER                                                batch/parameterized  90        running  2020-08-03T17:28:19Z
TRANSCRIPTOME_INDEX_i-0e7054c600dfc3dd3_12288                      batch/parameterized  65        running  2020-08-03T18:06:21Z
TRANSCRIPTOME_INDEX_i-0e7054c600dfc3dd3_16384                      batch/parameterized  65        running  2020-08-03T18:06:21Z
TRANSCRIPTOME_INDEX_i-0e7054c600dfc3dd3_2048                       batch/parameterized  65        running  2020-08-03T18:06:21Z
TRANSCRIPTOME_INDEX_i-0e7054c600dfc3dd3_3072                       batch/parameterized  65        running  2020-08-03T18:06:21Z
TRANSCRIPTOME_INDEX_i-0e7054c600dfc3dd3_32768                      batch/parameterized  65        running  2020-08-03T18:06:21Z
TRANSCRIPTOME_INDEX_i-0e7054c600dfc3dd3_4096                       batch/parameterized  65        running  2020-08-03T18:06:21Z
TRANSCRIPTOME_INDEX_i-0e7054c600dfc3dd3_65536                      batch/parameterized  65        running  2020-08-03T18:06:21Z
TRANSCRIPTOME_INDEX_i-0e7054c600dfc3dd3_8192                       batch/parameterized  65        running  2020-08-03T18:06:21Z
TXIMPORT                                                           batch/parameterized  50        running  2020-08-03T18:06:21Z
```
</details>

### CODE REVIEW

File: infrastructure/deploy.sh [hunk 1]
```diff
@@ -138,6 +138,11 @@ done
 # Copy ingress config to top level so it can be applied.
 cp deploy/ci_ingress.tf .
 
+# If this file still exists, the previous deploy didn't succeed, so recover
+if [ -f nomad-configuration/client-instance-user-data.tpl.sh.bak ]; then
+    mv nomad-configuration/client-instance-user-data.tpl.sh.bak nomad-configuration/client-instance-user-data.tpl.sh
+fi
```
Comment (kurtwheeler) [inline infrastructure/deploy.sh:144]: I'm confused by this. What case is this for? Why not just remake `nomad-configuration/client-instance-user-data.tpl.sh` anyway?

File: infrastructure/nomad-configuration/client-instance-user-data.tpl.sh [hunk 2]
```diff
@@ -59,10 +60,10 @@ export USER="${user}"
 #         let COUNTER=COUNTER+1
 # done
 
-# Only a single volume for now
-export EBS_VOLUME_INDEX=0
+# # Only a single volume for now
+# export EBS_VOLUME_INDEX=0
 
-sleep 25
+# sleep 25
```
Comment (kurtwheeler) [inline infrastructure/nomad-configuration/client-instance-user-data.tpl.sh:66]: Can we just remove these?

**Comment by wvauclain** [comment]
Leaving this here in case we need it in the future if our instance user data script gets too large:
1. We are encoding the gzipped tarball in `base64` since the output needs to be valid UTF-8, but alternate encodings like [ascii85](http://manpages.ubuntu.com/manpages/focal/man1/ascii85.1.html) would take up less space.
2. We are using `gzip`, but other compression algorithms like [xz](http://manpages.ubuntu.com/manpages/focal/man1/ascii85.1.html) have better compression ratios.

Both of these require installing additional packages though.

**Comment by wvauclain** [comment]
Before we merge this, I just need to take a look at the code and make sure that this does not break the dashboard on the frontend.

**Comment by wvauclain** [comment]
Alright, I took another look and we were treating volume indices as strings everywhere before, so we should be good.

**Pull request merged**